Make by-value struct passing work properly

vtjnash opened the pull request:
this is ~~a first cut at~~ (done) ~~rebasing #3466 onto master~~ (also done) ready-to-merge patch to improve ccall/cfunction compatibility. @Keno let me know if you see anything that looks wrong.
- [x] linux 32 (http://www.uclibc.org/docs/psABI-i386.pdf)
- [x] linux 64 (http://www.uclibc.org/docs/psABI-x86_64.pdf)
- [x] mac 64 (same as linux 64)
- [x] win 32 (https://msdn.microsoft.com/en-us/library/984x0h58.aspx)
- [x] win 64 (https://msdn.microsoft.com/en-us/library/zthk2dkh.aspx)

To keep track of issues (from comments) from the previous PR:
- [x] Need to add link to LDC
- [x] ~~Need to figure out what to do with fixed-size arrays.~~ (will be done later)
- [x] merge the Mutable module from Gtk into base (#9986)
- [x] implement https://gist.github.com/vtjnash/7296634 ~~(only cfunction updates remain)~~
- [x] build libccalltest on travis
- [x] update documentation on ccall/cfunction in manual
- [x] add documentation for Ref type to stdlib
- [x] more tests for cfunction/ccall/Ref interactions
- [x] use unsafe_load in cfunction when receiving isbits types
- [x] update docs again. add docs for cconvert / unsafe_convert
- [x] add notes to NEWS and email to julia-users list

future work:
- [ ] calling conventions for cfunction
- [x] #2818 (types as c-structs)
- [ ] emit cfunction stubs for any `jl_apply_generic` call (possible sketch of implementation : https://gist.github.com/vtjnash/019b4d5d63b64a772a31)
- [ ] more completely use tuples to emulate more stuff, like fixed size arrays
- [ ] change `cfunction(f, T, ...)` to automatically call `convert(T, ...)::T` on the return value of `f`.
- [ ] add `convert(::Type{T}, x::Ref{T})` method
- [ ] hide the typetag

~~NOTE: please DO NOT merge this until the cfunction API changes are finalized. (i don't want to need to update Gtk.jl twice)~~

zenna commented:
Any progress on this?

vtjnash commented:
Lots. It's currently blocked by a number of other PRs that are needed first however.

vtjnash committed:
fix linux calling conventions

vtjnash committed:
fix win32 and win64 calling conventions

vtjnash committed:
implement Ref{T} support in cfunction

vtjnash committed:
copy any isbits type we received in cfunction as a Ref{T}. this is a more precise reverse of what happens in ccall given an argument of type Ref{T}

vtjnash committed:
write calling-c-and-fortran-code manual section for new ccall/cfunction/Ref types, and improve error messages in code

vtjnash committed:
more docs, related to the ccall enhancements, for the Ref and Ptr Types

tkelman commented on test/runtests.jl:
```diff
@@ -10,7 +10,7 @@ testnames = [
     "sysinfo", "rounding", "ranges", "mod2pi", "euler", "show",
     "lineedit", "replcompletions", "repl", "replutil", "sets", "test", "goto",
-    "llvmcall", "grisu", "nullable", "meta", "profile",
+    "llvmcall", "grisu", "nullable", "meta", "profile", "ccall",
```
Won't this make `Base.runtests` fail in any binary install where `libccalltest` won't be present?

vtjnash replied:
my goal was to make libccalltest also available in the binary install

tkelman replied:
works for me

vtjnash commented:
i686 crash is perhaps a calling convention issue with mingw32 (http://www.angelcode.com/dev/callconv/callconv.html, https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36834)

vtjnash committed:
proper ABI implementation support for ccall, and start work on the same for cfunction

vtjnash committed:
print to stderr during ccalltest

this makes it possible to see how far the test got if it fails partway
through. this also is better under certain combinations of windows
consoles at actually printing the output.

vtjnash commented:
if this gets a green light from travis, then it's ready to merge

is there some way to notify all package maintainers that use `cfunction` of the massive API breakage that this will likely cause for them?

JeffBezanson commented:
Best would be to write a brief cheat sheet giving some examples of common changes that will be needed. That can be posted here and to julia-dev.

vtjnash committed:
doc: folders don't make good Makefile targets, since they can (and do) get sometimes touched unexpected, resulting in unncessary rebuilding

vtjnash committed:
doc: update documentation of unsafe functions, add docs for cconvert/unsafe_convert, add NEWS for cfunction breaking changes

ivarne commented:
I have a folder with the master branch of all METADATA packages. Will I find these issues greping for `cfunction`?

vtjnash commented:
most likely, yes

ivarne commented:
This is the list of results https://gist.github.com/ivarne/114503137611069535fe

tknopp commented:
Is this an API change something that can be handled in a compatible way between 0.3 and master? Lets please ensure that 0.3 packages do not break (see https://github.com/JuliaLang/Graphics.jl/issues/1#issuecomment-77353103 ) on my opinion about that)

IainNZ commented:
Basically everything in JuliaOpt using callbacks in some way, as @ivarne 's list shows.

IainNZ commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,551 @@ memory to be passed to the callee and filled in. Allocation of memory
-                      (Ptr{Int32}, Ptr{Float64}, Ptr{Int32}, Ptr{Float64}, Ptr{Int32}),
-                      &n, DX, &incx, DY, &incy)
-      return product
+Finally, you can use ``cfunction`` to actually generate a call to the
```
This reads really weird, like a paragraph is missing

vtjnash replied:
noted. (this paragraph mirrored the section in ccall that I derived it from, but the ccall one actually had paragraphs in between)

tkelman commented on src/init.c:
```diff
@@ -1263,11 +1263,17 @@ void jl_get_builtin_hooks(void)
     jl_memory_exception    = jl_new_struct_uninit((jl_datatype_t*)core("MemoryError"));
+    #ifdef _OS_LINUX_
+    #ifdef _P32
+        jl_float64_type->alignment = 4;
```
does this apply to win32?

vtjnash replied:
no

tkelman replied:
Thanks, just checking. I saw the win32 complex erf failure, scrolled through the diff, and noticed this before I got to the test files and saw those were being skipped before.

Thanks for working on the docs here, this PR is quite a bit better about documentation than a lot of other comparable changes to internals have been.

vtjnash replied:
@Keno why is this here? https://github.com/JuliaLang/julia/commit/4a697d145ab719c11b019264d9f579c48b03b258

MAX_ALIGN should be 4 anyways on `_OS_LINUX_` with `_P32`

Keno replied:
Not sure if it's different now, but back then the alignment would be set to 8, since it's an 8 byte object but on Linux 32 bit the ABI specifies 4 byte alignment.

tkelman commented on test/math.jl:
```diff
@@ -75,14 +75,13 @@ end
-    @test_approx_eq dawson(1+2im) -13.388927316482919244-11.828715103889593303im
-end
+
+@test_approx_eq erf(1+2im) -0.53664356577856503399-5.0491437034470346695im
```
should these be disabled on win32 for now, if it needs to wait until an llvm upgrade to pass there?
```
exception on 2: ERROR: LoadError: assertion failed: |erf(1 + 2im) - -0.536643565778565 - 5.049143703447035im| <= 8.881784197001252e-12
  erf(1 + 2im) = 9.588980733155008e-152 - 8.973262701968788e-29im
  -0.536643565778565 - 5.049143703447035im = -0.536643565778565 - 5.049143703447035im
  difference = 5.077581949584898 > 8.881784197001252e-12
 in error at error.jl:20
 in test_approx_eq at test.jl:137
 in anonymous at test.jl:145
 in test_approx_eq at test.jl:147
 in runtests at C:\projects\julia\test\testdefs.jl:78
 in anonymous at multi.jl:833
while loading math.jl, in expression starting on line 79
```

vtjnash replied:
yes, this will work after an llvm upgrade. not sure if i can hack around it until then though

tkelman commented on LICENSE.md:
```diff
@@ -70,6 +70,7 @@ The following components of Julia's standard library have separate licenses:
 Julia builds the following libraries by default, but does not use them itself:
 
 - [RMATH](http://www.r-project.org/Licenses/)
+- [LDC](https://github.com/ldc-developers/ldc/blob/master/LICENSE)
```
this should probably be in the first section since the `abi_*` sources are getting built into libjulia

vtjnash replied:
good catch. when this was first written the file wasn't sorted, so rebase hadn't noticed that everything else moved

vtjnash replied:
i'm not really sure where this should go however. none of the existing sections really apply to the idea that we just took part of their code, but don't actually use any part of LDC.

tkelman replied:
and I did the same with MUSL for getopt.c but it's in the "external" section

nalimilan replied:
> i'm not really sure where this should go however. none of the existing sections really apply to the idea that we just took part of their code, but don't actually use any part of LDC.

I'd create a new section just below `The following components of Julia's standard library have separate licenses:`. Also, it would be nice to give more precisions, as the linked file gives several different licenses depending on the parts of the code. If the exceptions do not apply to the parts that were adapted, it would make the interpretation of the Julia license as a whole much easier.

tkelman commented:
> my goal was to make libccalltest also available in the binary install

still needs to be done here, to not break `Base.runtests` in binaries

vtjnash commented:
> Is this an API change something that can be handled in a compatible way between 0.3 and master? Lets please ensure that 0.3 packages do not break (see JuliaLang/Graphics.jl#1 (comment) ) on my opinion about that)

seems unlikely.

> This is the list of results https://gist.github.com/ivarne/114503137611069535fe

thanks. most of those will be OK. anything that was previously a bitstype will be unchanged. anything else was changed in an incompatible way and now needs to be wrapped in `Ref{...}` to get the same behavior

vtjnash commented:
> > my goal was to make libccalltest also available in the binary install

> still needs to be done here, to not break Base.runtests in binaries

Travis works. Am I missing something?

tkelman commented:
> Travis works. Am I missing something?

I might be. I haven't built this branch (which apparently now has a conflict somewhere), but where does libccalltest get copied over during `make install` or `make dist`? Travis does the former but not the latter.

vtjnash commented:
the `test` directory gets copied wholesale. we can revisit during v0.4 packaging if we discover it's missing

vtjnash committed:
minor doc cleanup

vtjnash committed:
workaround a llvm33/mingw32 calling convention mismatch

timholy commented:
This is a heroic change, Jameson, and I'm impressed with the extent of the documentation.

IainNZ commented:
:+1: to that

Keno commented:
From me too! Thanks for taking it up and pushing it through @vtjnash.

nalimilan commented on src/abi_llvm.cpp:
```diff
@@ -0,0 +1,38 @@
+//
+//                         LDC – the LLVM D compiler
+//
+// This file is distributed under the BSD-style LDC license. See the LICENSE
```
That LICENSE file is not shipped with Julia AFAICT. Also, "BSD-style" is not very specific. If that's not the pure BSD (and which version of the BSD license?), probably better copy the license text in the header.

vtjnash replied:
this was copied verbatim from the LDC source file. you're probably right that we should update it to contain the relevant text from their LICENSE file: https://github.com/ldc-developers/ldc/blob/908815256ffe72b04c76309aa0cdbd439f81faed/LICENSE#L32-L58

kmsquire commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
-    ccall(:foo, Void, (Ptr{Cint}, Ptr{Cfloat}), width, range)
+    width = Ref{Cint}(0)
+    range = Ref{Cfloat}(0)
+    ccall(:foo, Void, (Ref{Cint}, Ref{Cfloat}), width, range)
```
Wondering if 
```julia
ccall(:foo, Void, (Ref{Cint}, Ref{Cfloat}), Ref{Cint}(width), Ref{Cfloat}(range))
```
 is legal?

vtjnash replied:
you can't make a `Ref{Cint}` that contains a `Ref{Cint}`, if that's what you're asking – the type's don't match

kmsquire replied:
Okay, spelling it out slightly more:

```julia
width = Cint(0)
range = Cfloat(3.14)
ccall(:foo, Void, (Ref{Cint}, Ref{Cfloat}), Ref{Cint}(width), Ref{Cfloat}(range))
```

vtjnash replied:
the explicit typecasts and construction of the `Ref{Cint}` objects are a bit verbose and unnecessary in that case, but yes, it will work

JeffBezanson replied:
Is `Ref{T}` in the C signature equivalent to `Ptr{T}`?

JeffBezanson replied:
Also, can this call be written as `ccall(:foo, Void, (Ref{Cint}, Ref{Cfloat}), 0, 0)`? We don't want people to think the extra verbosity is necessary.

stevengj replied:
Also, how do you get the data back *out* of the `Ref{T}` type?  `convert(T, x)` or `T(x)`, presumably?

vtjnash replied:
They support getindex/setindex "[]" style access (for commonality with arrays). Convert (call) syntax looks good too, to add.

stevengj replied:
okay, added a todo at the top.

vtjnash committed:
fix #10444

vtjnash committed:
copy LDC license directly into abi implementation files

johnmyleswhite commented:
Awesome job improving the ccall docs, Jameson.

JeffBezanson commented on base/boot.jl:
```diff
@@ -125,9 +125,10 @@ export
-    Bool, FloatingPoint, Float16, Float32, Float64, Number, Integer, Int, Int8, Int16,
-    Int32, Int64, Int128, Ref, Ptr, Real, Signed, UInt, UInt8, UInt16, UInt32,
-    UInt64, UInt128, Unsigned,
+    Real, Complex, Number, Integer, Bool, Ref, Ptr,
```
`Complex` shouldn't be here; it's not in this module.

vtjnash replied:
i wonder if it should move here (or had at some points in the life of this PR), since it is now an influential type on ccall ABI generation

JeffBezanson replied:
Not a big deal; the run time system already refers to some stuff in Base.

JeffBezanson commented on doc/stdlib/c.rst:
```diff
@@ -209,10 +262,35 @@
+   data region of that type.
+
+   The ``Ref`` type is useful for creating garbage-collector safe pointers and
+   returning values from a function (esp. a c-function), for example.
```
Does "c-function" mean a function written in C, or a julia function passed to C with `cfunction`?

vtjnash replied:
either reading is correct, although "function written in C" was probably the intended reading

StefanKarpinski commented on LICENSE.md:
```diff
@@ -27,6 +27,10 @@ for exceptions.
+Julia includes code from the following projects, which have their own licenses:
+- [LDC](https://github.com/ldc-developers/ldc/blob/master/LICENSE) (for ccall/cfunction ABI definitions)
+- [MUSL](http://git.musl-libc.org/cgit/musl/tree/COPYRIGHT) (for getopt implementations on Windows)
+- [NetBSD](http://www.netbsd.org/about/redistribution.html) (for setjmp/longjmp implementations on Windows)
```
The LDC license concerns me a bit. It's not a common license and it's not simple either. The main MUSL license is MIT, same as Julia; is the getopt implementation under that license or a different one? Similarly, is the NetBSD setjmp/longjmp code on Windows under the 2 or 3 clause BSD license? The 2-clause is MIT-equivalent, while the 3-clause is slightly more restrictive (but not much).

Keno replied:
The applicable license from LDC is 3-clause BSD since it's part of LDC, but not DMD.

StefanKarpinski replied:
Ah, that's good. We should probably note these things here. We may want to go through all of the code we use from other projects at some point and check what the specific license of the code we're using is (as opposed to the whole project). I think we're actually already in pretty good shape regarding this.

vtjnash replied:
it's already there at the top of each file that we've taken from another project

tkelman replied:
The getopt implementation in musl doesn't have a specific license header, so it's under the main musl license. I copied that main musl license into the file: https://github.com/JuliaLang/julia/blob/master/ui/getopt.c

nalimilan replied:
If the MUSL/getopt.c license is MIT, then please don't list it here: it's the same as the rest of Julia. Things are already complex enough regarding files which have a different license.

Not sure for NetBSD, but if that's exactly equivalent to MIT, maybe do not list it here either.

tkelman replied:
> it's the same as the rest of Julia

No it isn't. Musl is copyright Rich Felker, Julia isn't.

nalimilan replied:
Yes, but that information isn't very useful, except for people owning the copyright for most of the rest of the Julia codebase and who may want to relicense it (which likely isn't going to happen). Is it really needed here, rather than in the file header?

Anyway, if it's equivalent to MIT, it should be written explicitly.

StefanKarpinski replied:
2-clause BSD is equivalent to MIT but not identical; it should be listed as a separate license.

JeffBezanson commented on NEWS.md:
```diff
@@ -247,6 +253,36 @@ Deprecated or removed
+Improvements to ccall and cfunction, and Ref
+--------------------------------------------
+
+  * a `Ref` abstract type was added, as a general replacement for the `&` special-syntax.
```
Needs to be clearer. Does this mean "instead of writing `&x`, write `Ref(x)`"?

vtjnash replied:
either `Ref(x)` or `Ref{T}`

StefanKarpinski commented:
This is epic. Especially love the wonderfully detailed documentation.

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
-any expression, such as ``&0`` or ``&f(x)``.
+Notice that we have to be careful about the return type: ``qsort`` expects a function
+returning a C ``int``, so we must be sure to return ``Cint`` via a call to ``convert``
+and a ``typeassert``.
```
(I wish that `cfunction` would just call `convert` and add the typeassert for you.)

vtjnash replied:
Can you edit the top post and add it as future work? It wouldn't actually be that hard to add anymore.

stevengj replied:
Done.

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+:Complex64: Exactly corresponds to the ``complex float`` type in C (or ``COMPLEX*8`` in Fortran)
+:Complex128: Exactly corresponds to the ``complex double`` type in C (or ``COMPLEX*16`` in Fortran)
+:Signed: Exactly corresponds to the ``signed`` type annotation in C (or any ``INTEGER`` type in Fortran). Any Julia type that is not a subtype of ``Signed`` is assumed to be unsigned
+:Ref{T}: Behaves like a ``Ptr{T}`` that owns it's memory
```
it's -> its

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+    `cconvert_gcroot` function will attempt to first make a null-terminated copy of the array with
+    each element replaced by it's ``cconvert`` version. This allows, for example, passing an ``argv``
+    pointer array of type ``Vector{ByteString}`` to an argument of type ``Ptr{Ptr{Cchar}}``.
 
```
This is confusing to me: a `Vector{ByteString}` is not an array of eltype `Ptr{Cchar}`, so the text seems inconsistent.

vtjnash replied:
Correct. Hence the comment about it making a copy  The name of the function called out in this help text is wrong (from an earlier version of the PR)

stevengj replied:
The sentence begins "If an array of eltype `Ptr{T}` is passed as a `Ptr{Ptr{T}}` argument".  Grammatically, that means that the business about making the copy only applies to this case.   Hence, it does not apply to passing "``Vector{ByteString}`` to an argument of type ``Ptr{Ptr{Cchar}}``".

Maybe you meant to write: "If an array whose eltype can be converted to `Ptr{T}` is passed as a `Ptr{Ptr{T}}` argument" ... ?

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
 it can be passed to C functions expecting NUL-terminated data without making
 a copy.
+`Note`: C functions that take an argument of the type ``char**`` can be called by using
+a ``Ptr{Ptr{UInt8}}`` type within Julia. For example,
```
Shouldn't that be `Ptr{Ptr{Cchar}}` for consistency?

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+Memory allocation and deallocation of such objects must be
+handled by calls to the appropriate cleanup routines in the libraries
+being used, just like in any C program. Do not try to free an object
+received from a C library with `c_free` in Julia, as this may result
```
What do you do if a C library returns a pointer and says "deallocate this pointer with `free` when you are done with it"...?  That's not uncommon.

mlubin replied:
On windows at least it's unsafe in general to call free() across shared library boundaries.

vtjnash replied:
Correct. Technically that is a bug in their library (Unless that library happened to be called libc...)

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+types can be used to recursively mirror c-style structs,
+without requiring as much manual management of the ``Ptr`` conversions.
+After #2818 is implemented, it will be true that an `Vector{T}` will be equivalent to
+an `Ptr{Ptr{T}}`. That is currently not true, and the conversion must be explicitly.
```
must be performed explicitly.

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+Use ``Ptr{T}`` if the memory is expected to be populated by ``C`` (without type-tag).
+Use ``Ref{T}`` if you have an ``isbits`` type,
+but you want to turn it into a pointer to a struct in another struct definition.
+
```
Isn't `Ref{T}` also used for passing by reference, not just in structs?   In general, this paragraph is confusing and seems to use some esoteric terminology (type-tag?).

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+
+  + depends on how this parameter is used, first translate this to the intended pointer type,
+    then determine the Julia equivalent using the remaining rules in this list
+  + this argument may be declared as ``Ptr{Void}``, if it really is just an unknown pointer
```
unknown pointer -> opaque pointer

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+
+  + ``Ref{T}``, where ``T`` is the Julia type corresponding to ``T``
+  + argument value will be copied if it is an ``isbits`` type
+    otherwise, the value must be a valid Julia object
```
What is an "invalid Julia object"?

vtjnash replied:
Memory that wasn't allocated by Julia's object allocator

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+* ``jl_value_t*``
+
+  + ``Any``
+  + argument value must be a valid Julia object
```
argument value?  I thought this was about return types...

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+
+* ``(T*)(...)`` (e.g. a pointer to a function)
+
+  + ``Ptr{Void}`` (you may need to use ``cfunction`` explicitly to create this pointer)
```
the comment about `cfunction` seems out-of-place for return values

stevengj commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+Special Reference Syntax for ccall (deprecated):
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    The ``&`` syntax is deprecated, use the ``Ref{T}`` argument type instead
```
Why not just make `&(x::T)` syntactic sugar for `Ref{T}(x)`, rather than deprecating it?  The `Ref` syntax would be a lot cleaner that way.

JeffBezanson replied:
Yes, we should pick one way and use it consistently. The cleanest syntax seems to be `ccall(f, R, (Ref{T},), x)`. Even fewer characters than it takes now, and I believe it handles all uses of `&`.

You only need to manually construct a `Ref` object if you want to get data back from C, or if you want to point to an element within an array. And in those cases you can still use `Ref{T}` as the argument type.

vtjnash replied:
I agree with Jeff

stevengj replied:
An even cleaner syntax would be `ccall(f, R, (&T,), x)`, with `Ref` objects constructed by `&x` if necessary to get data back from C.  Then most users would never have to deal with `Ref{T}`.

vtjnash replied:
Not compatible with cfunction. Also requires specialized parsing (sorry brief -- iPhone is having trouble typing on this page)

vtjnash replied:
scratch that. i just tried it, and it would potentially actually require *removing* the special parsing of `&` to make that possible. then you could define:
```
&(A) = Ref(A)
&(A::Type) = Ref{A}
```
but i don't really recommend abusing and replacing the bitwise-and operator in this way.

i went with a type-based approach precisely because it avoided needing special-syntax to define ccall (which would have transferred over to ccall). It should be easier now to work with the AST (since there aren't `Expr(:&)` nodes to deal with) and easier to automatically generate ccall/cfunction expressions (which also was the motivation for attempting to minimize the number of places that codegen branches on isbits/bitstype/leaftype, rather than syntactic expressions).

StefanKarpinski replied:
+1 to no special syntax for this


> On Mar 10, 2015, at 7:22 PM, Jameson Nash <notifications@github.com> wrote:
> 
> In doc/manual/calling-c-and-fortran-code.rst:
> 
> >  
> >  This is used extensively in Julia's LAPACK interface, where an integer ``info``
> >  is passed to LAPACK by reference, and on return, includes the success code.
> >  
> > +Special Reference Syntax for ccall (deprecated):
> > +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> > +
> > +    The ``&`` syntax is deprecated, use the ``Ref{T}`` argument type instead
> scratch that. i just tried it, and it would potentially actually require removing the special parsing of & to make that possible. then you could define:
> 
> &(A) = Ref(A)
> &(A::Type) = Ref{A}
> but i don't really recommend abusing and replacing the bitwise-and operator in this way.
> 
> i went with a type-based approach precisely because it avoided needing special-syntax to define ccall (which would have transferred over to ccall). It should be easier now to work with the AST (since there aren't Expr(:&) nodes to deal with) and easier to automatically generate ccall/cfunction expressions (which also was the motivation for attempting to minimize the number of places that codegen branches on isbits/bitstype/leaftype, rather than syntactic expressions).
> 
> —
> Reply to this email directly or view it on GitHub.
>

blakejohnson commented on doc/manual/calling-c-and-fortran-code.rst:
```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
+    typedef returntype (*functiontype)(argumenttype,...)
+
+The function `cfunction` generates the c-compatible function pointer for
+a call to a Julia Julia library function.
```
A double "Julia" slipped in here.

zenna commented:
Typos?

`It is possible to pass Julia functions to native c-functions that accept function pointer arguments`

`The function cfunction generates the c-compatible function pointer for a call to a Julia Julia library function. Arguments to cfunction are as follows:`

pao commented:
@zenna fixed in 36ccc5b9e2cfd34b5bb702e4de9441b2fa63a1d6; thanks!
